File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/Broadcast.java
package com.github.dedis.popstellar.model.network.method;

import androidx.annotation.NonNull;

import com.github.dedis.popstellar.model.Immutable;
import com.github.dedis.popstellar.model.network.method.message.MessageGeneral;
import com.github.dedis.popstellar.model.objects.Channel;

import java.util.Objects;

/**
 * Broadcast a high level message inside a container.
 *
 * <p>Does not expect any answer
 */
@Immutable
public final class Broadcast extends Message {

  private final MessageGeneral message;

  /**
   * Constructor for a Broadcast
   *
   * @param channel name of the channel
   * @param message the message to broadcast
   * @throws IllegalArgumentException if any parameter is null
   */
  public Broadcast(Channel channel, MessageGeneral message) {
    super(channel);
    if (message == null) {
      throw new IllegalArgumentException("Trying to broadcast a null message");
    }
    this.message = message;
  }

  @Override
  public String getMethod() {
    return Method.MESSAGE.getMethod();
  }

  /** Returns the message of the Broadcast. */
  public MessageGeneral getMessage() {
    return message;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    if (!super.equals(o)) {
      return false;
    }
    Broadcast that = (Broadcast) o;
    return Objects.equals(getMessage(), that.getMessage());
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), getMessage());
  }

  @NonNull
  @Override
  public String toString() {
    return "Broadcast{"
        + "channel='"
        + getChannel()
        + "', method='"
        + getMethod()
        + "', message="
        + getMessage()
        + "}";
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/Broadcast.kt
package com.github.dedis.popstellar.model.network.method

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.model.network.method.message.MessageGeneral
import com.github.dedis.popstellar.model.objects.Channel
import java.util.Objects

/**
 * Broadcast a high level message inside a container.
 *
 * Does not expect any answer
 */
@Immutable
class Broadcast(channel: Channel?, message: MessageGeneral?) : Message(channel) {
  /** Returns the message of the Broadcast. */
  val message: MessageGeneral

  /**
   * Constructor for a Broadcast
   *
   * @param channel name of the channel
   * @param message the message to broadcast
   * @throws IllegalArgumentException if any parameter is null
   */
  init {
    requireNotNull(message) { "Trying to broadcast a null message" }

    this.message = message
  }

  override val method: String
    get() = Method.MESSAGE.method

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    if (!super.equals(other)) {
      return false
    }
    val that = other as Broadcast
    return message == that.message
  }

  override fun hashCode(): Int {
    return Objects.hash(super.hashCode(), message)
  }

  override fun toString(): String {
    return "Broadcast{channel='$channel', method='$method', message=$message}"
  }
}
